src/enclave/enclave_oe.c

```diff
@@ -311,38 +305,117 @@ static void _read_eeid_config()
     sgxlkl_enclave_state.config = cfg;
 }
 
+void sgxlkl_ensure_inside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_within_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is not in enclave memory.\n");
+}
+
+void sgxlkl_ensure_outside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_outside_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is in enclave memory.\n");
+}
+
 static void _copy_shared_memory(const sgxlkl_shared_memory_t* host)
 {
     const sgxlkl_enclave_config_t* cfg = sgxlkl_enclave_state.config;
 
-    /* Deep copy where necessary */
+    /* The host shared memory struct has been copied by OE (but not its
+     * contents). */
+    sgxlkl_ensure_inside(host, sizeof(sgxlkl_shared_memory_t));
 
-    sgxlkl_shared_memory_t* enc = &sgxlkl_enclave_state.shared_memory;
-    memset(enc, 0, sizeof(sgxlkl_shared_memory_t));
+    /* Temporary, volatile copy to make sure checks aren't reordered */
+    volatile sgxlkl_shared_memory_t* enc = oe_calloc_or_die(
+        1,
+        sizeof(sgxlkl_shared_memory_t),
+        "Could not allocate memory for shared memory\n");
 
     if (cfg->io.network)
+    {
         enc->virtio_net_dev_mem = host->virtio_net_dev_mem;
+        sgxlkl_ensure_outside(
+            enc->virtio_net_dev_mem, sizeof(struct virtio_dev));
+    }
 
     if (cfg->io.console)
+    {
         enc->virtio_console_mem = host->virtio_console_mem;
+        sgxlkl_ensure_outside(
+            enc->virtio_console_mem, sizeof(struct virtio_dev));
+    }
 
     enc->evt_channel_num = host->evt_channel_num;
-    /* enc_dev_config is required to be outside the enclave */
-    enc->enc_dev_config = host->enc_dev_config;
+
+    if (enc->evt_channel_num > 0)
+    {
+        _Static_assert(
+            sizeof(enc_dev_config_t) == 24UL,
+            "enc_dev_config_t size has changed");
+
+        enc->enc_dev_config = oe_calloc_or_die(
+            enc->evt_channel_num,
+            sizeof(enc_dev_config_t),
+            "Could not allocate memory for event channel device config\n");
+
+        for (size_t i = 0; i < enc->evt_channel_num; i++)
+        {
+            const enc_dev_config_t* host_conf_i = &host->enc_dev_config[i];
+            enc_dev_config_t* enc_conf_i = &enc->enc_dev_config[i];
+
+            enc_conf_i->dev_id = host_conf_i->dev_id;
+
+            enc_evt_channel_t* host_chn = host_conf_i->enc_evt_chn;
+            if (host_chn)
+            {
+                enc_conf_i->enc_evt_chn = host_chn;
```

Comment:
@prp when I try to deep-copy `enc_evt_chn` (see commented block), the devices don't work anymore (execution hangs). So, the host must be changing that pointer (plus possibly the contents of the `enc_evt_channel_t`). Is this expected or does this indicate that something somewhere must be going wrong?

Reply:
The host shouldn't be changing the pointer to the event channels, but event channels are shared 64-bit values between enclave and guest and both will need to read / write the values.

Reply:
When I allocate a new `enc_evt_chn_t` and copy the contents from the host:

    enc_conf_i->enc_evt_chn = oe_calloc_or_die(...);
    *enc_conf_i->enc_evt_chn = *host_chn;

Then the devices don't work anymore, so I suspect the host does indeed change the pointer at some point. I'll dig a bit deeper.

Reply:
I think this is copying one too deep.  The `enc_evt_channel_t*` is a pointer to a shared 64-bit value.  If you copy the initial value, you no longer have an event channel, you just have an event.